driver-core/src/test/java/com/datastax/driver/core/PreparedStatementMultiNodeTest.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (C) 2012-2017 DataStax Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import com.datastax.driver.core.exceptions.DriverInternalError;
+import com.datastax.driver.core.utils.CassandraVersion;
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.fail;
+
+@CCMConfig(numberOfNodes = 2, dirtiesContext = true)
+@CreateCCM(CreateCCM.TestMode.PER_METHOD)
+public class PreparedStatementMultiNodeTest extends CCMTestsSupport {
+
+    private static final String keyspace2 = TestUtils.generateIdentifier("ks_");
+    private static final String keyspace3 = TestUtils.generateIdentifier("ks_");
+
+    @Override
+    public void onTestContextInitialized() {
+        execute(
+                String.format(TestUtils.CREATE_KEYSPACE_SIMPLE_FORMAT, keyspace2, 2),
+                String.format("CREATE TABLE %s.users(id int, id2 int, name text, primary key (id, id2))", keyspace2),
+                String.format("INSERT INTO %s.users(id, id2, name) VALUES (2, 2, 'test2')", keyspace2),
+                String.format(TestUtils.CREATE_KEYSPACE_SIMPLE_FORMAT, keyspace3, 2),
+                String.format("CREATE TABLE %s.users(id int, id2 int, name text, primary key (id, id2))", keyspace3),
+                String.format("INSERT INTO %s.users(id, id2, name) VALUES (3, 3, 'test3')", keyspace3)
+        );
+    }
+
+    @Test(groups = "long")
+    @CassandraVersion("4.0.0")
+    public void should_be_able_to_execute_statement_on_restarted_node_reprepare_on_up_set_keyspace() {
```

Comment:
The method names are really hard to understand. "be able to" can generally be removed without losing anything. Also, the fact that reprepare on up is enabled or not is a detail, especially with recent Cassandra version where it will make no difference; the two cases could be done in the same method.

So my suggestion would be:
```java
    @Test(groups = "long")
    @CassandraVersion("4.0.0")
    public void should_execute_statement_with_per_query_keyspace_on_restarted_node() {
        executeAfterNodeBroughtBackUp(true, true);
        executeAfterNodeBroughtBackUp(false, true);
    }

    @Test(groups = "long")
    public void should_execute_statement_with_session_keyspace_on_restarted_node() {
        executeAfterNodeBroughtBackUp(true, false);
        executeAfterNodeBroughtBackUp(false, false);
    }
```